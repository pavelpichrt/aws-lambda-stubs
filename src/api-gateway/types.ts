import {
  APIGatewayEventDefaultAuthorizerContext,
  APIGatewayEventRequestContextWithAuthorizer,
  APIGatewayProxyEvent,
  APIGatewayTokenAuthorizerEvent,
  Context,
} from 'aws-lambda';

export interface APIGatewayAuthorizer {
  principalId: string;
  integrationLatency: number;
}

export interface APIGatewayEventContextParams {
  resourceId?: string;
  authorizer?: APIGatewayAuthorizer;
  resourcePath: string;
  path: string;
  stage: string;
  apiId: string;
  userAgent: string;
  sourceIp: string;
  httpMethod?: string;
}

export interface APIGatewayEventParams {
  resourceId?: string;
  authorizer?: APIGatewayAuthorizer;
  stage?: string;
  resourcePath?: string;
  queryStringParameters?: Record<string, string>;
  apiId?: string;
  userAgent?: string;
  sourceIp?: string;
  pathParameters?: Record<string, string>;
  body?: object;
  headers?: Record<string, string>;
  includeAuthorizationHeader?: boolean;
  authorizationHeaderGenerator?: () => string;
  path?: string;
  httpMethod?: string;
}

export interface APIGatewayRequestContextParams {
  functionName?: string;
  memoryLimitInMB?: number;
}

export interface APIGatewayEventContextGenerator {
  ({
    resourceId,
    authorizer,
    resourcePath,
    path,
    stage,
    apiId,
    userAgent,
    sourceIp,
    httpMethod,
  }: APIGatewayEventContextParams): APIGatewayEventRequestContextWithAuthorizer<APIGatewayEventDefaultAuthorizerContext>;
}
export interface APIGatewayEventGenerator {
  ({
    resourceId,
    authorizer,
    stage,
    resourcePath,
    queryStringParameters,
    apiId,
    userAgent,
    sourceIp,
    pathParameters,
    body,
    headers,
    includeAuthorizationHeader,
    authorizationHeaderGenerator,
    path,
    httpMethod,
  }?: APIGatewayEventParams): APIGatewayProxyEvent;
}

export interface APIGatewayContextGenerator {
  ({ functionName, memoryLimitInMB }?: APIGatewayRequestContextParams): Context;
}

export interface APIGatewayAuthorizerEventParams {
  authorizationToken?: string;
  resource?: string;
}

export interface APIGatewayAuthorizerEventGenerator {
  ({
    authorizationToken,
    resource,
  }?: APIGatewayAuthorizerEventParams): APIGatewayTokenAuthorizerEvent;
}
