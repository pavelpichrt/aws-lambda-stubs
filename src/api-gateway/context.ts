import { v4 as uuidv4 } from 'uuid';

import { randomAlphaNumericString } from '../support';
import { APIGatewayContextGenerator } from './types';

export const apiGatewayProxyContext: APIGatewayContextGenerator = ({
  functionName,
  memoryLimitInMB,
} = {}) => {
  const actualFunctionName = functionName || 'mock-function-name';
  const functionVersion = '1';
  const now = new Date();

  return {
    callbackWaitsForEmptyEventLoop: true,
    functionVersion,
    functionName: actualFunctionName,
    memoryLimitInMB: memoryLimitInMB ? memoryLimitInMB.toString() : '128',
    logGroupName: `/aws/lambda/${actualFunctionName}`,
    logStreamName: `${now.getFullYear()}/${
      now.getMonth() + 1
    }/${now.getDate()}/[${functionVersion}]${randomAlphaNumericString(32)}`,
    invokedFunctionArn: `arn:aws:lambda:eu-west-1:123456278901:function:${actualFunctionName}:latest`,
    awsRequestId: uuidv4(),
    getRemainingTimeInMillis: () => now.getTime() + 10000 * 60, // 10 minutes from now
    done: () => null,
    fail: () => null,
    succeed: () => null,
  };
};
