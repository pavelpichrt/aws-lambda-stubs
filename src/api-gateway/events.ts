import { v4 as uuidv4 } from 'uuid';

import { generateRandomInt, randomAlphaNumericString } from '../support';
import {
  APIGatewayAuthorizerEventGenerator,
  APIGatewayEventContextGenerator,
  APIGatewayEventGenerator,
} from './types';

const generateEventContext: APIGatewayEventContextGenerator = ({
  resourceId,
  authorizer,
  resourcePath,
  path,
  stage,
  apiId,
  userAgent,
  sourceIp,
  httpMethod = 'POST',
}) => {
  const actualAuthorizer = authorizer || {
    principalId: 'mock-principal',
    integrationLatency: generateRandomInt(5, 500),
  };
  const monthNames = [
    'Jan',
    'Feb',
    'Mar',
    'Apr',
    'May',
    'Jun',
    'Jul',
    'Aug',
    'Sep',
    'Oct',
    'Nov',
    'Dec',
  ];
  const now = new Date();

  return {
    resourceId: resourceId || randomAlphaNumericString(6),
    authorizer: actualAuthorizer,
    resourcePath,
    httpMethod,
    extendedRequestId: `${randomAlphaNumericString(15)}=`,
    requestTime: `${now.getDate()}/${
      monthNames[now.getMonth()]
    }/${now.getFullYear()}:${now.getHours()}:${now.getMinutes()}:${now.getSeconds()} +0000`,
    path,
    accountId: '123456278901',
    protocol: 'HTTP/1.1',
    stage,
    domainPrefix: apiId,
    requestTimeEpoch: now.getTime(),
    requestId: uuidv4(),
    identity: {
      cognitoIdentityPoolId: null,
      apiKey: null,
      apiKeyId: null,
      clientCert: null,
      accountId: null,
      cognitoIdentityId: null,
      caller: null,
      sourceIp,
      principalOrgId: null,
      accessKey: null,
      cognitoAuthenticationType: null,
      cognitoAuthenticationProvider: null,
      userArn: null,
      userAgent,
      user: null,
    },
    domainName: `${apiId}.execute-api.eu-west-2.amazonaws.com`,
    apiId,
  };
};

export const apiGatewayProxyEvent: APIGatewayEventGenerator = ({
  resourceId,
  authorizer,
  stage = 'mock-stage',
  resourcePath = '/mock/{id}/path',
  apiId = 'mockapi',
  userAgent = 'MockApi/1.0.0',
  sourceIp = '123.123.123.123',
  pathParameters,
  queryStringParameters = null,
  body,
  headers = {
    'Content-Type': 'application/json',
  },
  authorizationHeaderGenerator,
  includeAuthorizationHeader = true,
  path = `/${stage}/${resourcePath}`,
  httpMethod = 'POST',
} = {}) => {
  const actualPathParameters: Record<string, string> = {};
  (resourcePath.match(/({[a-zA-Z0-9_-]+})/g) || []).forEach(
    (fullParam: string) => {
      const paramWithoutBrackets = fullParam.substring(1, fullParam.length - 1);
      const paramValue = pathParameters?.[paramWithoutBrackets] || 'foo';

      path = path.replace(fullParam, paramValue);
      actualPathParameters[paramWithoutBrackets] = paramValue;
    },
  );
  const additionalHeaders: Record<string, string> = headers;

  if (includeAuthorizationHeader) {
    const defaultHeaderGenerator = () => randomAlphaNumericString(32);
    const actualAuthorizationHeaderGenerator =
      authorizationHeaderGenerator || defaultHeaderGenerator;

    additionalHeaders.Authorization = actualAuthorizationHeaderGenerator();
  }

  const eventContext = generateEventContext({
    resourceId,
    authorizer,
    resourcePath,
    path,
    stage,
    apiId,
    userAgent,
    sourceIp,
    httpMethod,
  });
  const traceId = `Root=1-${randomAlphaNumericString(
    8,
  )}-${randomAlphaNumericString(24)}`;
  const additionalHeadersMultiValue: Record<string, string[]> = {};

  for (const [headerName, headerValue] of Object.entries(additionalHeaders)) {
    additionalHeadersMultiValue[headerName] = [headerValue];
  }

  return {
    resource: resourcePath,
    path,
    httpMethod,
    queryStringParameters,
    headers: {
      Accept: '*/*',
      'Accept-Encoding': 'gzip, deflate, br',
      Host: `${apiId}.execute-api.eu-west-2.amazonaws.com`,
      'User-Agent': userAgent,
      'X-Amzn-Trace-Id': traceId,
      'X-Forwarded-For': sourceIp,
      'X-Forwarded-Port': '443',
      'X-Forwarded-Proto': 'https',
      ...additionalHeaders,
    },
    multiValueHeaders: {
      Accept: ['*/*'],
      'Accept-Encoding': ['gzip, deflate, br'],
      Host: [`${apiId}.execute-api.eu-west-2.amazonaws.com`],
      'User-Agent': [userAgent],
      'X-Amzn-Trace-Id': [traceId],
      'X-Forwarded-For': [sourceIp],
      'X-Forwarded-Port': ['443'],
      'X-Forwarded-Proto': ['https'],
      ...additionalHeadersMultiValue,
    },
    multiValueQueryStringParameters: null,
    pathParameters: actualPathParameters,
    stageVariables: null,
    requestContext: eventContext,
    body: body ? JSON.stringify(body) : null,
    isBase64Encoded: false,
  };
};

export const apiGatewayAuthorizerEvent: APIGatewayAuthorizerEventGenerator = ({
  authorizationToken = 'authorization-token-stub',
  resource = '/prod/POST/{proxy+}',
} = {}) => ({
  type: 'TOKEN',
  authorizationToken,
  methodArn: `arn:aws:execute-api:eu-west-1:123456789012:example${resource}`,
});
