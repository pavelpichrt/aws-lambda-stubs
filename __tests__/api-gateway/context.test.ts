import { Context } from 'aws-lambda';
import { apiGatewayProxyContext } from '../../src/api-gateway/context';
import { uuidRegex } from '../config';

const logStreamRegex = /^\d{4}\/\d{2}\/\d{2}\/\[\d+\][a-z0-9]{32}$/;

const validateContext = (context: Context) => {
  const { logStreamName, awsRequestId, ...deterministicParams } = context;

  expect(deterministicParams).toMatchSnapshot();
  expect(logStreamName).toEqual(expect.stringMatching(logStreamRegex));
  expect(awsRequestId).toEqual(expect.stringMatching(uuidRegex));
};

describe('apiGatewayProxyContext', () => {
  it('Generates a correct stub with default arguments', () => {
    const actualContext = apiGatewayProxyContext();

    validateContext(actualContext);
  });

  it('Generates a correct stub with arguments specified', () => {
    const actualContext = apiGatewayProxyContext({
      functionName: 'my-function',
      memoryLimitInMB: 512,
    });

    validateContext(actualContext);
  });
});
