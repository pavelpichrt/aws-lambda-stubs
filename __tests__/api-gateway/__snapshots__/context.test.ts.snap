// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiGatewayProxyContext Generates a correct stub with arguments specified 1`] = `
{
  "callbackWaitsForEmptyEventLoop": true,
  "done": [Function],
  "fail": [Function],
  "functionName": "my-function",
  "functionVersion": "1",
  "getRemainingTimeInMillis": [Function],
  "invokedFunctionArn": "arn:aws:lambda:eu-west-1:123456278901:function:my-function:latest",
  "logGroupName": "/aws/lambda/my-function",
  "memoryLimitInMB": "512",
  "succeed": [Function],
}
`;

exports[`apiGatewayProxyContext Generates a correct stub with default arguments 1`] = `
{
  "callbackWaitsForEmptyEventLoop": true,
  "done": [Function],
  "fail": [Function],
  "functionName": "mock-function-name",
  "functionVersion": "1",
  "getRemainingTimeInMillis": [Function],
  "invokedFunctionArn": "arn:aws:lambda:eu-west-1:123456278901:function:mock-function-name:latest",
  "logGroupName": "/aws/lambda/mock-function-name",
  "memoryLimitInMB": "128",
  "succeed": [Function],
}
`;
